### DISCUSSION TITLE
Psrpoperator: passing a dynamic variable

### DISCUSSION THREAD

**Discussion by akblackwel** [apache/airflow#36931] [Q&A]
I was curious what I need to do to pass psrpoperator a dynamic variable.

It can't pick up xcom
params isn't templated

In a pythonoperator I can get dag environment with Variable.get

I'm at a loss to make the psrpoperator environment aware

I just don't see a way to pass in any airflow variables

Thanks

**Comment by potiuk** [comment]
You can change template_fields as you want:

1) you can extend the operator and make the field templated or even in your dag set

```python
class MyPSRPOperator(PsrpOperator):
    template_fields =  tuple(*PsrpOperator.template_fields, 'field_1', 'field2')
```

and use such custom operator

2) You can manipulate template_fields on class level in your DAG 

```python
PsrpOperator.template_fields  = tuple(*PsrpOperator.template_fields, 'field_1', 'field2')
```

**Comment by akblackwel** [comment]
Hmmm, 

I've missed this in the documentation, but after looking at your post, I was able to find this documented in the airflow site.

[https://airflow.apache.org/docs/apache-airflow/stable/howto/custom-operator.html#add-template-fields-with-subclassing](url)

I tried both.

using tuple, seem to fail

using the documented way, I'm getting an error based on the following code.

I always get some form of the error below.

Broken DAG: [/air_osp/dags/ocsp/dag_osp_osp_test_psrp.py] Traceback (most recent call last):
  File "/airflow/python3/lib/python3.7/site-packages/airflow/models/baseoperator.py", line 429, in apply_defaults
    result = func(self, **kwargs, default_args=default_args)
  File "/airflow/python3/lib/python3.7/site-packages/airflow/models/baseoperator.py", line 789, in __init__
    f"Invalid arguments were passed to {self.__class__.__name__} (task_id: {task_id}). "
airflow.exceptions.AirflowException: Invalid arguments were passed to MyPsrpOperator (task_id: TaskID1). Invalid arguments were:
**kwargs: {'argu': {'test'}}


```python
from typing import Sequence

with DAG(
     {...}
     params={"Environment":"DEV1"},
     {...}

class MyPsrpOperator(PsrpOperator):
    template_fields: Sequence[str] = (*PsrpOperator.template_fields, "argu")
    def __init__(self, name: str, **kwargs) -> None:
        super().__init__(**kwargs)
        self.name = name
py_op_prsp_02 = MyPsrpOperator(
    task_id="TaskID1",
    psrp_conn_id="psrp_conn", 
    powershell=taskid1,
    wsman_options={"ssl": False, "auth": "ntlm"},
    do_xcom_push=False,
    params={"username": username, "password": password},
    argu={"test"} # other options tried argu={"name": {{params.Environment}}"}
)

```

**Reply by potiuk** [reply]
Well. you have to have `argu` field defined in your operator - if you don't it won't work.

**Reply by potiuk** [reply]
And to explain: template_fields does not refer to parameters of the constructor, they refer to fields in the operator - so if you want custom field, then well, you have to have that field.
